lib/tree.rb

```diff
@@ -16,40 +16,129 @@ def initialize
+  #           current = current.right
+  #         else
+  #           current.right = TreeNode.new(key, value)
+  #           return
+  #         end
+  #       end
+  #     end
+  #   end
+  # end
+
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+
+    return nil if current.nil?
+
+    until current.key == key
+      if current.key > key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+    return current.value
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list) #l, n, r
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value} # root is getting pushed to the list 
+    inorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) n represents the number of nodes in a tree
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list) #n, l, r
+    return list if current_node.nil?
+
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list) #l, r, n
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+    return list
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def height_helper(current_node)
+    return 0 if current_node.nil?
+
+    left_depth = height_helper(current_node.left)
+    right_depth = height_helper(current_node.right)
+
+    if left_depth > right_depth
+      return left_depth + 1
+    else
+      return right_depth + 1
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def height
```

Comment:
Note that this is a recursive method so you're going to have space complexity due to the call stack.